File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline


def is_possible(row, col):
    for r in range(row):
        # 같은 열에 행이 있거나, 같은 대각선에 퀸이 있으면 False
        if graph[r] == col or abs(row - r) == abs(col - graph[r]):
            return False
    return True


def backtracking(row):
    global result

    # base
    if row == n:
        result += 1
        return

    for col in range(n):
        if is_possible(row, col):
            graph[row] = col  # 가능하면 해당 행의 열에 퀸 놓기
            backtracking(row + 1)  # 다음 행으로 이동


n = int(input())
graph = [0] * n  # 같은 행에 어차피 놓일 수 없으므로 2차원 리스트 필요 없음

result = 0
backtracking(0)
print(result)

# 퀸이 서로 공격할 수 없게 놓아야 함
# N * N 체스판에 (0, 0)부터 퀸을 놓아보고 공격이 가능할 경우 백트래킹
# 같은 행, 열, 대각에 위치하면 안됨

"""
0 0 0
0 0 0
0 0 0

(1, 1): (0, 0) - (2, 2) - (2, 0) - (0, 2)
row(행): 1
col(열): 1

(1, 1)에 퀸이 있다고 가정

0, 0: abs(1 - 0) == abs(1 - 0) -> 1
2, 2: abs(1 - 2) == abs(1 - 2) -> 1
2, 0: abs(1 - 2) == abs(1 - 0) -> 1
0, 2: abs(1 - 0) == abs(1 - 2) -> 1
"""

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;
import java.util.*;

public class Main {
    static int n;
    static int result = 0;
    static int[] queens;

    public static boolean check(int current_row, int current_col) {
        for (int prev_row = 0; prev_row < current_row; prev_row++) {
            if (queens[prev_row] == current_col || Math.abs(current_row - prev_row) == Math.abs(current_col - queens[prev_row]))
            return false;
        }
        return true;

        // current_row: 1, current_col: 0 ~ 3
        // 이전 퀸: (2, 3)
        // 현재 퀸: (4, 5)
        // 행 차이: abs(4 - 2) = 2
        // 열 차이: abs(5 - 3) = 2
        // Math.abs(current_row - prev_row) -> 놓으려는 행과 이전 퀸의 행의 차이
        // Math.abs(current_col - queens[prev_row]) -> 놓으려는 열과 이전 퀸의 열의 차이


        /*
         * (1, 1)을 기준으로 대각
         * (0, 0) = abs(1 - 0) == abs(1 - 0) = 1
         * (2, 2) = abs(1 - 2) == abs(1 - 2) = 1
         * (0, 2) = abs(1 - 0) == abs(1 - 2) = 1
         */
    }

    public static void backtracking(int row) {
        // base
        if (row == n) {
            result += 1;
            return;
        }
        
        for (int col = 0; col < n; col++) {
            if (check(row, col)) {
                queens[row] = col;
                backtracking(row + 1);
            } 
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        n = Integer.parseInt(br.readLine());
        queens = new int[n];

        backtracking(0);

        bw.write(result + "\n");
        bw.flush();
    }
}
